onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+          }
+        }
+      }
+
+      fused_outputs.insert(fused_outputs_to_add.begin(), fused_outputs_to_add.end());
+
+      // Sort inputs and outputs by the order they were added
+      std::multimap<int, const NodeArg*> inputs, outputs;
+
+      for (auto it = fused_inputs.begin(); it != fused_inputs.end(); ++it) {
+        inputs.insert(std::pair<int, const NodeArg*>(it->second, it->first));
+      }
+
+      for (auto it = fused_outputs.begin(); it != fused_outputs.end(); ++it) {
+        outputs.insert(std::pair<int, const NodeArg*>(it->second, it->first));
+      }
+
+      // Assign inputs and outputs to subgraph's meta_def
+      auto meta_def = std::make_unique<::onnxruntime::IndexedSubGraph::MetaDef>();
+      meta_def->name = "TRTKernel";
+      meta_def->domain = kMSDomain;
+
+      for (const auto& input : inputs) {
+        meta_def->inputs.push_back(input.second->Name());
+      }
+
+      for (const auto& output : outputs) {
+        meta_def->outputs.push_back(output.second->Name());
+      }
+
+      meta_def->since_version = 1;
+      sub_graph->SetMetaDef(meta_def);
+    }
+
+    result.push_back(std::make_unique<ComputeCapability>(std::move(sub_graph)));
+  }
+
+  return result;
+}
+
+std::shared_ptr<KernelRegistry> TRTExecutionProvider::GetKernelRegistry() const {
+  return nullptr;
+}
+
+common::Status TRTExecutionProvider::CopyTensor(const Tensor& src, Tensor& dst) const {
+  ORT_UNUSED_PARAMETER(src);
+  ORT_UNUSED_PARAMETER(dst);
+  return Status::OK();
+}
+
+common::Status TRTExecutionProvider::Compile(const std::vector<onnxruntime::Node*>& fused_nodes,
+                                             std::vector<NodeComputeInfo>& node_compute_funcs) {
+  for (const auto* fused_node : fused_nodes) {
+    std::unordered_map<std::string, int> input_map;
+    std::vector<int> input_indexes;
+    std::vector<int> input_binding_indexes;
+    std::vector<int> input_dim_sizes;
+    std::vector<int> output_binding_indexes;
+    std::vector<int> output_dim_sizes;
+    std::vector<std::vector<int64_t>> output_shapes;
+    std::vector<std::string> graph_input_name;
+    std::vector<std::vector<int>> input_shape;
+
+    // Build map from input name to its index in InputDefs
+    for (int i = 0, end = fused_node->InputDefs().size(); i < end; ++i) {
+      input_map[fused_node->InputDefs()[i]->Name()] = i;
+    }
+
+    const auto* func_body = fused_node->GetFunctionBody();
+    if (!func_body) {
+      return common::Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT, "Function body is empty");
+    }
+
+    // Reconstruct graph from fused node's function body
+    const Graph& graph_body = fused_node->GetFunctionBody()->Body();
+    onnxruntime::Model model(graph_body.Name(), true, ModelMetaData(), IOnnxRuntimeOpSchemaRegistryList(), graph_body.DomainToVersionMap());
+    onnxruntime::Graph& graph = model.MainGraph();
+
+    for (const auto& graph_body_node : graph_body.Nodes()) {
+      graph.AddNode(graph_body_node);
```

Comment:
not sure i'm following this. why do we need to add the nodes back into the graph after we had previously already fused it?

Reply:
nevermind. i understand now. TRT expects the modelproto with full graph. so we have to reconstruct it.

Reply:
Your are right.

Reply:
Would it be a lot simpler for us if they could take a GraphProto? Or do they need something from the ModelProto level?

---
In reply to: [261488525](https://github.com/Microsoft/onnxruntime/pull/382#discussion_r261488525) [](ancestors = 261488525)

Reply:
TRT parser needs serialized ModelProto as its input. We've talked about using onnxruntime IR format as the inputs, which could save the back and forth conversions, but there is no agreement yet.